src/lib/mongoBackend/mongoRegistrationCreate.cpp

```diff
@@ -0,0 +1,219 @@
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Ken Zangelin
+*/
+#include <string>
+#include <vector>
+
+#include "mongo/client/dbclient.h"
+
+#include "logMsg/logMsg.h"
+#include "logMsg/traceLevels.h"
+
+#include "common/sem.h"
+#include "common/statistics.h"
+#include "common/errorMessages.h"
+#include "rest/OrionError.h"
+#include "rest/HttpStatusCode.h"
+#include "apiTypesV2/Registration.h"
+#include "mongoBackend/dbConstants.h"
+#include "mongoBackend/safeMongo.h"
+#include "mongoBackend/MongoGlobal.h"
+#include "mongoBackend/connectionOperations.h"
+#include "mongoBackend/mongoRegistrationCreate.h"
+
+
+
+/* ****************************************************************************
+*
+* setRegistrationId - 
+*/
+static void setRegistrationId(mongo::BSONObjBuilder* bobP, std::string* regIdP)
+{
+  mongo::OID oId;
+
+  oId.init();
+
+  *regIdP = oId.toString();
+  bobP->append("_id", oId);
+}
+
+
+
+/* ****************************************************************************
+*
+* setDescription -
+*/
+static void setDescription(const std::string& description, mongo::BSONObjBuilder* bobP)
+{
+  if (description != "")
+  {
+    bobP->append(REG_DESCRIPTION, description);
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* setExpiration -
+*/
+static void setExpiration(long long expires, mongo::BSONObjBuilder* bobP)
+{
+  if (expires != -1)
+  {
+    bobP->append(REG_EXPIRATION, expires);
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* setServicePath -
+*/
+static void setServicePath(const std::string& servicePath, mongo::BSONObjBuilder* bobP)
+{
+  if (servicePath != "")
+  {
+    bobP->append(REG_SERVICE_PATH, servicePath);
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* setContextRegistrationVector - 
+*/
+static void setContextRegistrationVector(ngsiv2::Registration* regP, mongo::BSONObjBuilder* bobP)
+{
+  mongo::BSONArrayBuilder  contextRegistration;
+  mongo::BSONArrayBuilder  entities;
+  mongo::BSONArrayBuilder  attrs;
+
+  for (unsigned int eIx = 0; eIx < regP->dataProvided.entities.size(); ++eIx)
+  {
+    ngsiv2::EntID* eP = &regP->dataProvided.entities[eIx];
+    std::string    eId;
+    std::string    eType;
+
+    eId           = (eP->idPattern   != "")? eP->idPattern : eP->id;
+    eType         = (eP->typePattern != "")? eP->typePattern : eP->type;
+
+    //
+    // FIXME PR: should patterns be supported or not?
```

Comment:
Not fully sure (I'd need to check how subscription works to be sure) but I'd say that patterns are not supported by the moment in registrations. Do we have an issue about pattern supports in registrations?

Reply:
I found none ...
The question is, do we want this?

Reply:
Of course, it would be great to have it :) But it is very difficult to implement it completly. (Deja vu mode on) If the request to be forwared includes a pattern and the registration also is based on patterns we have the problem of knowing if the set defined by a regex is included in antoher regex, which its a problem with a not general solution as far as I know (deja vu mode off :).

Thus, let's keep this simple in the first implementation, so this is not yet supported.

In fact, this is already explained in our documentation (for NGSIv1 case), see https://fiware-orion.readthedocs.io/en/master/user/context_providers/index.html:

> * In NGSIv1 registrations, `isPattern` cannot be set to `"true"`. If so, the registration fails and an error is returned. The OMA specification allows for regular expressions in entity id in registrations but as of now, the Context Broker doesn't support this feature.

Reply:
ok

Reply:
Fixed in 5686ca2